doc/developer/design/20210630_linearizability.md

```diff
@@ -0,0 +1,108 @@
+The two functions `get_read_ts` and `get_write_ts` provide this.
+Those functions do not need to change because they are sometimes called outside of SQL transactions, and we only need to provide linearizability for user-observed events.
+
+For example,  the `send_builtin_table_updates_at_offset` function calls `get_write_ts` when logging tables are being updated.
+But since that's an internal table update and a user did not observe the write as having happened (which is the case for an INSERT), we don't need to use the HashMap.
+
+This linearizability system can thus be viewed as a layer above the current timestamp determination functions, timelines, and time domains, since it specifically only cares about timestamps that users initiated.
+
+### Wall Time Going Backward
+
+Concerning only to the realtime timeline: the system clock may go backward (due to, say, some time synchronization program like NTP setting the clock).
+This will result in no new data will be produced from (rather, we will not advance the frontiers of) realtime sources.
+SELECTs will continue to function normally, but without any updates until the system clock catches up to what it was.
+
+Tables also have a side effect that if reads and writes are interleaved, the writes will increase the timestamp by at least 1 (if the system time did not increase).
+Thus in worst-case scenarios table writes would be observed, but no new data from realtime sources, until the system clock catches up.
+
+### Restarts
+
+Assume some bad case of materialize shutting down, the system clock going backward a lot, then materialize restarting.
+On restart, the source durability timestamps are read from the catalog.
+Sources will not respond to queries until they have processed the previously processed data which will be assigned the timestamps from the persisted durability timestamps.
+
+### Non-Persisted Tables and Volatile Sources
+
+Tables today are not persisted across restarts.
+Some sources are volatile and not guaranteed to have the same data in them at all times.
+If materialize restarts, we only guarantee linearizability for non-volatile data with durable timestamps.
+
+## Alternatives
+
+Instead of advancing timestamps on each transaction, we use index `upper`s to advance the timestamp-per-timeline.
+When `update_upper` is run, it would bump the timeline's timestamp to the max of what is there and the new upper.
+This would potentially cause read queries to block if another index in the same timeline is noticably ahead of another.
+
+## Open questions
+
+Will this cause regular blocking for realtime sources if tables are involved a lot because those always bump the ts, but realtime sources might take up to 1s to catch up to that?
+
+`determine_timestamp` and `determine_frontier` do not consult source durability frontiers.
```

Comment:
What are these functions, and are they germane to the design for coordinator linearizability?

Reply:
You're right. Not needed. Removed.